### ISSUE TITLE
Generated builders do not generate @Override annotations when appropriate

### ISSUE THREAD

**Issue by tjwilson90** [immutables/immutables#202]
Example:
```java
@Value.Immutable
public interface Foo {
    long getBar();
    interface Builder {
        Builder bar();
    }
}
```
Result:
The `bar()` method in the generated `ImmutableFoo.Builder` class will not have an `@Override` annotation.

**Comment by elucash** [comment]
Yep, this part is not covered, we might analyze the methods we extends, but this will definitely will add some runtime overhead and complexity. Generated files have `@SuppressWarning` unless it is disabled `Style.generateSuppressAllWarnings = false`. I wonder which actual errors/warning you are getting.
Another option is to extend generated classes (reverse extends) as shown http://immutables.github.io/immutable.html#extending-builder, but this might not be what you need as both ways have it's pros and cons.
Let us know you if it's important issue, could you work around it? Thank you for the issue report!

**Comment by tjwilson90** [comment]
The issue is that the overrides are missing from the Builder class which does not have @SuppressWarnings("all") on it. Simply adding that would be fine for me.

**Comment by tjwilson90** [comment]
Also, reverse extends appears to be broken in 2.1.0 (Builder class is final).

**Comment by elucash** [comment]
> class which does not have @SuppressWarnings("all")

That's strange. Have you disabled `generateSuppressAllWarnings=false` in your styles (because it's turned on by default)? Can you share small repro-case?

> Also, reverse extends appears to be broken in 2.1.0 (Builder class is final).

You're not the first who got this, but this functionality is tested on CI, so it might be more completed then just don't work. Try to recompile, I've heard it might be due some caching issue. If the problem remains,  please, provide some repro-case, it's definitely need to be investigated and fixed

Thank you for the information!

**Comment by tjwilson90** [comment]
Sure, put this in a bash script and execute
```
#!/bin/bash

mkdir repro
cd repro
cat << EOF > Foo.java
import org.immutables.value.Value;

@Value.Immutable
public interface Foo {
    long getBar();
    interface Builder {
        Builder bar();
    }
}
EOF

wget http://search.maven.org/remotecontent?filepath=org/immutables/value/2.1.0/value-2.1.0.jar -O value-2.1.0.jar
wget http://search.maven.org/remotecontent?filepath=org/immutables/value/2.1.0/value-2.1.0-annotations.jar -O value-2.1.0-annotations.jar

javac -classpath value-2.1.0.jar:value-2.1.0-annotations.jar Foo.java
```
It will fail to compile. The resulting ImmutableFoo.Builder class will not have @SuppressWarnings("all") and it will be final.

**Comment by tjwilson90** [comment]
err, that should be 
```
interface Builder {
    Builder bar(long bar);
}
```
which will get it to compile, but the two issues still remain.

**Comment by elucash** [comment]
Thank you for providing a repro script.
Surprisingly I do see `SuppressWarning("all")` generated at the top of `ImmutableFoo`. So maybe it has no effect in your environment somehow. Any hints?

On the other hand, you could properly extend the generated builder, because when you actually do this, annotation processor detects this fact and no longer generates it as `final`. But then attribute initializer bar(long) is indeed generated as final, you cannot override it: we were going to relax this, but that causes some conceptual consistency problems for different kind of overloads, at least we are not ready allow them to be overridden, but in extension `Builder` you can add alternative overloads or additional initializers which might call original ones.

Here's slightly augmented version of the script I've used
```
#!/bin/bash

rm -fr repro 
mkdir repro
cd repro
cat << EOF > Foo.java
import org.immutables.value.Value;

@Value.Immutable
public interface Foo {
    long getBar();
    class Builder extends ImmutableFoo.Builder {
        // Builder bar(long b);
    }
}
EOF

curl -L http://search.maven.org/remotecontent?filepath=org/immutables/value/2.1.0/value-2.1.0.jar > value-2.1.0.jar
# no need for value:annotations, those are only needed if annotation processing
# use different classpath than compile class path.
javac -classpath value-2.1.0.jar Foo.java
```
Here's files created by this script:
https://www.dropbox.com/s/kw9dq9ze0c28uc3/repro.zip?dl=1

**Comment by tjwilson90** [comment]
My problem is not with the @SuppressWarnings("all") on ImmutableFoo. That does show up as expected.
My problem is that @SuppressWarnings("all") is not put on ImmutableFoo.Builder.

**Comment by tjwilson90** [comment]
Reverse builder extension does appear to be working for me.

**Comment by elucash** [comment]
I thought that  `@SuppressWarnings("all")` should just cover all nested definitions. If nested builder class will have `@SuppressWarnings("all")` annotation, would it resolve the problem you have?

When I compile with `-Xlint:overrides`, I don't see any warning for the case

```java
@Value.Immutable
public interface Foo {
    long getBar();
    interface Builder {
        Builder bar(long bar);
    }
}
```
Which compiler produces such warning/error in this case?

**Comment by tjwilson90** [comment]
I get the warning when using ecj in eclipse.

**Comment by tjwilson90** [comment]
Huh, weird, it's actually fine if I configure missing overrides to be a warning in eclipse (don't get an error or a warning), but if I configure it to be an error, I get an error.

**Comment by tjwilson90** [comment]
```
#!/bin/bash

rm -fr repro 
mkdir repro
cd repro
cat << EOF > Foo.java
import org.immutables.value.Value;

@Value.Immutable
public interface Foo {
    long getBar();
    interface Builder {
        Builder bar(long bar);
    }
}
EOF

wget http://central.maven.org/maven2/org/eclipse/jdt/core/compiler/ecj/4.3.1/ecj-4.3.1.jar -O ecj-4.3.1.jar
wget http://search.maven.org/remotecontent?filepath=org/immutables/value/2.1.0/value-2.1.0.jar -O value-2.1.0.jar

java -jar ecj-4.3.1.jar -1.7 -err:+allOver-ann -classpath value-2.1.0.jar Foo.java
```

**Comment by tjwilson90** [comment]
hmm, but manually adding @SuppressWarnings("all") to the builder class and disabling annotation processing doesn't cause the error to be suppressed either. Apparently suppressing warnings can't suppress warnings that are being compiled as errors.

Ok, no need for a @SuppressWarnings annotation on the Builder class. Would still be nice to get proper override annotations on the Builder methods, but I understand if you decide not to do it for complexity/performance reasons.

**Comment by elucash** [comment]
Overall I support the idea of flagging of such missing overrides as errors, let this hang on a wish list, maybe we'll code something to analyze this, however priority is far from critical. Thank you for helping nail down the issue!

**Comment by vorburger** [comment]
@elucash this problem just bit me too (also using ecj, in CLI not IDE; have it conf. as error not warning).
